crypto/src/key/secp256k1/extended_keys.rs

```diff
@@ -0,0 +1,256 @@
+        // Create a new chain code
+        let mut chain_code = [0u8; 32];
+        rng.fill_bytes(&mut chain_code);
+        let chain_code = chain_code.into();
+        let private_key = secp256k1::SecretKey::new(rng).into();
+        // Generate a new private key
+        let ext_priv = Secp256k1ExtendedPrivateKey {
+            private_key,
+            chain_code,
+        };
+        // Generate the public key
+        let ext_pub = Secp256k1ExtendedPublicKey::from_private_key(&ext_priv);
+        // Return the pair
+        (ext_priv, ext_pub)
+    }
+
+    pub fn as_private_key(&self) -> &Secp256k1PrivateKey {
+        &self.private_key
+    }
+}
+
+impl Derivable for Secp256k1ExtendedPrivateKey {
+    fn derive_child(self, num: ChildNumber) -> Result<Self, DerivationError> {
+        let mut mac = HmacSha512::new_from_slice(&self.chain_code.into_array())
+            .expect("HMAC can take key of any size");
+
+        let secp_key = self.private_key.data;
+
+        // We only support hard derivations
+        if num.is_hardened() {
+            mac.update(&[0u8]);
+            mac.update(&secp_key[..]);
+        } else {
+            return Err(DerivationError::UnsupportedDerivationType);
+        }
+
+        // Add the child number
+        mac.update(&num.to_encoded_be_bytes());
+
+        let mut result = mac.finalize().into_bytes();
+        let private_key = secp256k1::SecretKey::from_slice(&result[..32])
+            .map_err(|_| DerivationError::KeyDerivationError)?
+            // TODO [SECURITY] a Scalar is created here but not erased
+            .add_tweak(&secp_key.into())
+            .map_err(|_| DerivationError::KeyDerivationError)?
+            .into();
+
+        let chain_code: [u8; CHAINCODE_LENGTH] =
+            result[32..].try_into().expect("Chaincode size is 32 bytes");
+        let chain_code = ChainCode::from(chain_code);
+
+        result.zeroize();
+
+        Ok(Secp256k1ExtendedPrivateKey {
+            private_key,
+            chain_code,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Encode, Decode)]
+pub struct Secp256k1ExtendedPublicKey {
+    /// Chain code
+    pub chain_code: ChainCode,
+    /// Public key
+    pub public_key: Secp256k1PublicKey,
+}
+
+impl Secp256k1ExtendedPublicKey {
+    pub fn as_public_key(&self) -> &Secp256k1PublicKey {
+        &self.public_key
+    }
+
+    pub fn from_private_key(private_key: &Secp256k1ExtendedPrivateKey) -> Self {
+        Secp256k1ExtendedPublicKey {
+            public_key: private_key.private_key.data.x_only_public_key(SECP256K1).0.into(),
+            chain_code: private_key.chain_code,
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::key::hdkd::derivation_path::DerivationPath;
+    use bip39::Mnemonic;
+    use hex::ToHex;
+    use serialization::DecodeAll;
+    use serialization::Encode;
+    use std::str::FromStr;
+
+    #[test]
+    fn test_serialization() {
```

Comment:
Nitpick: there isn't much sense to use the `test_` prefix.

Reply:
Good point, I changed it. I have searched and there are many tests that are `fn test_`